engine/access/state_stream/backend/backend_executiondata.go

```diff
@@ -32,26 +34,45 @@ type ExecutionDataBackend struct {
 
 	subscriptionHandler  *subscription.SubscriptionHandler
 	executionDataTracker tracker.ExecutionDataTracker
+
+	executionResultProvider optimistic_sync.ExecutionResultInfoProvider
+	executionStateCache     optimistic_sync.ExecutionStateCache
 }
 
-func (b *ExecutionDataBackend) GetExecutionDataByBlockID(ctx context.Context, blockID flow.Identifier) (*execution_data.BlockExecutionData, error) {
-	header, err := b.headers.ByBlockID(blockID)
+func (b *ExecutionDataBackend) GetExecutionDataByBlockID(
+	ctx context.Context,
+	blockID flow.Identifier,
+	criteria optimistic_sync.Criteria,
+) (*execution_data.BlockExecutionData, *accessmodel.ExecutorMetadata, error) {
+	execResultInfo, err := b.executionResultProvider.ExecutionResultInfo(blockID, criteria)
 	if err != nil {
-		return nil, fmt.Errorf("could not get block header for %s: %w", blockID, err)
+		return nil, nil, fmt.Errorf("failed to get execution result for block %v: %w", blockID, err)
```

Comment:
this should check for `storage.ErrNotFound` and convert to `codes.NotFound`. as it is now, this would return an internal error.